File: cpp_version/mixed_recall/linked_palindrome/linked_palindrome.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits.h>
#include <math.h>
#include <queue>
#include <stack>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// linked palindrome
// Write a function, linked_palindrome, that takes in the head of a linked list as an argument.
//  The function should return a boolean indicating whether
// or not the linked list is a palindrome. A palindrome is a sequence that is the same both forwards and backwards.

class Node {
public:
    int val;
    Node *next;

    Node(int initialVal) {
        val = initialVal;
        next = nullptr;
    }
};

bool linkedPalindrome(Node *head) {
    // todo

        return true;

}

int main() {

    Node a(3);
    Node b(2);
    Node c(7);
    Node d(7);
    Node e(2);
    Node f(3);

    a.next = &b;
    b.next = &c;
    c.next = &d;
    d.next = &e;
    e.next = &f;

    // 3 -> 2 -> 7 -> 7 -> 2 -> 3
    std::cout << std::boolalpha << linkedPalindrome(&a) << std::endl; // 1 (true)
    // Node a(3);
    // Node b(2);
    // Node c(4);
    c.val = 4;
    a.next = &b;
    b.next = &c;
    c.next = nullptr;
    // 3 -> 2 -> 4
    std::cout << linkedPalindrome(&a) << std::endl; // 0 (false)
    // Node a(3);
    // Node b(2);
    // Node c(3);
    c.val = 3;

    a.next = &b;
    b.next = &c;
    c.next = nullptr;

    // 3 -> 2 -> 3
    std::cout << linkedPalindrome(&a) << std::endl; // 1 (true)

    Node a1(0);
    Node b1(1);
    Node c1(0);
    Node d1(1);
    Node e1(0);

    a1.next = &b1;
    b1.next = &c1;
    c1.next = &d1;
    d1.next = &e1;

    // 0 -> 1 -> 0 -> 1 -> 0
    std::cout << linkedPalindrome(&a1) << std::endl; // 1 (true)
    // Node a(0);
    // Node b(1);
    // Node c(0);
    // Node d(1);
    // Node e(1);
    e1.val = 1;
    a1.next = &b1;
    b1.next = &c1;
    c1.next = &d1;
    d1.next = &e1;

    // 0 -> 1 -> 0 -> 1 -> 1
    std::cout << linkedPalindrome(&a1) << std::endl; // 0 (false)
    Node a2(5);

    // 5
    std::cout << linkedPalindrome(&a2) << std::endl;     // 1 (true)
    std::cout << linkedPalindrome(nullptr) << std::endl; // 1 (true)

    return 0;
}

File: python_version/mixed_recall/linked_palindrome/linked_palindrome.py
# linked palindrome
# Write a function, linked_palindrome, that takes in the head of a linked list as an argument.
#  The function should return a boolean indicating whether
# or not the linked list is a palindrome. A palindrome is a sequence that is the same both forwards and backwards.


class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


def linked_palindrome(head):

    return


a = Node(3)
b = Node(2)
c = Node(7)
d = Node(7)
e = Node(2)
f = Node(3)

a.next = b
b.next = c
c.next = d
d.next = e
e.next = f

# 3 -> 2 -> 7 -> 7 -> 2 -> 3
print(linked_palindrome(a))  # True
a = Node(3)
b = Node(2)
c = Node(4)

a.next = b
b.next = c

# 3 -> 2 -> 4
print(linked_palindrome(a))  # False
a = Node(3)
b = Node(2)
c = Node(3)

a.next = b
b.next = c

# 3 -> 2 -> 3
print(linked_palindrome(a))  # True
a = Node(0)
b = Node(1)
c = Node(0)
d = Node(1)
e = Node(0)

a.next = b
b.next = c
c.next = d
d.next = e

# 0 -> 1 -> 0 -> 1 -> 0
print(linked_palindrome(a))  # True
a = Node(0)
b = Node(1)
c = Node(0)
d = Node(1)
e = Node(1)

a.next = b
b.next = c
c.next = d
d.next = e

# 0 -> 1 -> 0 -> 1 -> 1
print(linked_palindrome(a))  # False
a = Node(5)

# 5
print(linked_palindrome(a))  # True
print(linked_palindrome(None))  # True
